tempesta_fw/http_parser.c

```diff
@@ -786,84 +674,112 @@ do {									\
 		 */							\
 		r = TFW_POSTPONE;					\
 		__fsm_const_state = to;					\
+		if (tfw_http_msg_add_data_ptr(msg, &msg->body, data, len)) \
+			return TFW_BLOCK;				\
 		goto done;						\
 	}								\
 	c = *p;								\
 	goto to;							\
 } while (0)
 
-#define __FSM_B_MOVE(to)	__FSM_B_MOVE_n(to, 1)
+#define __FSM_B_MOVE(to)						\
+	__FSM_B_MOVE_n(to, 1)
 
-#define TFW_HTTP_INIT_BODY_PARSING(msg, to_state)			\
+#define TFW_HTTP_INIT_BODY_PARSING(msg)					\
 do {									\
 	TFW_DBG3("parse msg body: flags=%#x content_length=%lu\n",	\
 		 msg->flags, msg->content_length);			\
 	/* RFC 2616 4.4: firstly check chunked transfer encoding. */	\
 	if (msg->flags & TFW_HTTP_CHUNKED)				\
-		__FSM_B_MOVE(to_state);					\
+		__FSM_MOVE_nofixup(RGen_Body);				\
 	/* Next we check content length. */				\
 	if (msg->content_length						\
 	    && !(msg->flags & TFW_HTTP_VOID_BODY))			\
 	{								\
 		parser->to_read = msg->content_length;			\
-		__FSM_B_MOVE(to_state);					\
+		__FSM_MOVE_nofixup(RGen_Body);				\
 	}								\
 	/* There is no body at all. */					\
+	msg->body.flags |= TFW_STR_COMPLETE;				\
 	r = TFW_PASS;							\
 	FSM_EXIT();							\
 } while (0)
 
 #define TFW_HTTP_PARSE_BODY()						\
 /* Read request|response body. */					\
 __FSM_STATE(RGen_Body) {						\
+	tfw_http_msg_set_data(msg, &msg->body, p);			\
+	/* fall through */						\
+}									\
+__FSM_STATE(RGen_BodyChunk) {						\
 	TFW_DBG3("read body: to_read=%d\n", parser->to_read);		\
 	if (!parser->to_read) {						\
-		__fsm_sz = __data_remain(p);				\
-		/* Read next chunk length. */				\
-		__fsm_n = parse_int_hex(p, __fsm_sz, &parser->_tmp.acc);\
-		TFW_DBG3("len=%zu ret=%d to_read=%lu\n",		\
-			 __fsm_sz, __fsm_n, parser->_tmp.acc);		\
-		switch (__fsm_n) {					\
-		case CSTR_POSTPONE:					\
-			/* Not all data has been parsed. */		\
-			__FSM_B_MOVE_n(RGen_Body, __fsm_sz);		\
-		case CSTR_BADLEN: /* bad header length */		\
-		case CSTR_NEQ: /* bad header value */			\
+		/* prevent @parse_int_hex false positives */		\
```

Comment:
Not a big deal, but it's better to follow normal English grammar in comments: start sentence from Capital letter and finish it with dot. However, short trailing comments which are "part of the sentence", like `case CSTR_NEQ: /* bad header value */\` above can bypass the rules. Just a small note to make code nice.